catboost/private/libs/algo/leafwise_scoring.cpp

```diff
@@ -341,23 +342,43 @@ static void CalcScoresForSubCandidate(
             updateSplitScoreClosure);
     };
 
-    // TODO(ilyzhin) make caching for nonsymmetric trees
-    if (!ctx->UseTreeLevelCaching() || ctx->Params.ObliviousTreeOptions->GrowPolicy != EGrowPolicy::SymmetricTree) {
-        TVector<TBucketStats> stats;
-        stats.yresize(bucketCount);
+    TArrayRef<TBucketStats> statsRef = statsForSubtractionTrick.GetStatsRef();
+    TArrayRef<TBucketStats> parentStatsRef = statsForSubtractionTrick.GetParentStatsRef();
+    TArrayRef<TBucketStats> siblingStatsRef = statsForSubtractionTrick.GetSiblingStatsRef();
 
+    auto calcStatsScores = [&] (TArrayRef<TBucketStats> stats) {
+        const bool useSubtractionTrick = parentStatsRef.data() != nullptr && siblingStatsRef.data() != nullptr;
```

Comment:
&& approxDimension == 1

after that please remove isMultiClass and IsMultiClass from GreedyTensorSearchDepthwise